// api-sdk/index.js
/*
	Initialization Script
	requires all other private modules needed for the API SDK to work
*/
var $http = require('./modules/http.js'),
	$model = require('./modules/model.js'),
	$login = require('./modules/login.js'),
	$renew = require('./modules/renew.js');
/*

SAMPLE CONFIG

var config = {
	host:"api.meditab.com",
	port: 80,
	credentials:{
  		id: "553f7b0dbd4e648e02e24f63",
  		key: "P9BmWUtlUgomte-zQtS-HjAWj6CydZTXpQmxeXwI"
	}
}

*/
var sdk = {
	initialize: function(configuration,callback){
		var credentials = configuration.credentials;
		var message = "api sdk initialized succesfully";

		if(credentials && credentials.id && credentials.key){
			
			process.MEDITAB = (function(){
				var private_config = {
					host: "api.meditab.com",
					port: 80,
					credentials: false
				}

				for(var key in configuration){
					private_config[key] = configuration[key];
				}
				return function(){
					return private_config;
				}
			}());
			
			console.log(message);
			if(callback){
				callback({
					success: message
				});
			}
		}else{
			message = "api credential id and key not configured";
			console.log(message);
			if(callback){
				callback({
					error: message
				});
			}
		}

	}
};

sdk.http = $http;
sdk.model = $model;
sdk.login = $login;
sdk.renew = $renew;

module.exports = sdk;
