// api-sdk/modules/model.js
/*
Use Case:

var user = new AO("appuser"); //argument is a model name in the api

*/

var REQUEST = require('./request.js');

var AO; // ApiObject

AO = function(name) {
  return this.name = name;
};

AO.prototype.get = function(url) {
  var req;
  req = new REQUEST(url, "get");
  return req;
};

AO.prototype.post = function(url, data) {
  var req;
  req = new REQUEST(url, "post");
  if (data) {
    req.data = data;
  }
  return req;
};

AO.prototype.put = function(url, data) {
  var req;
  req = new REQUEST(url, "put");
  if (data) {
    req.data = data;
  }
  return req;
};

AO.prototype.del = function(url) {
  var req;
  req = new REQUEST(url, "delete");
  return req;
};

AO.prototype.show = function(id) {
  var url;
  url = this.name + "/show/" + id;
  return this.get(url);
};

AO.prototype.list = function() {
  var url;
  url = this.name + "/list";
  return this.get(url);
};


AO.prototype.find = function(query) {
  var req, url;
  url = this.name + "/search";
  req = this.get(url);
  //req.params.push("q=" + apiUtils.serialize(query));
  req.params.push({
  	key:"q",
  	value: encodeURIComponent(JSON.stringify(query))
  })
  return req;
};

AO.prototype.create = function(data) {
  var url;
  url = this.name + "/create";
  return this.post(url, data);
};

AO.prototype.update = function(id, data) {
  var url;
  url = this.name + "/update/" + id;
  return this.put(url, data);
};

AO.prototype["delete"] = function(id) {
  var url;
  url = this.name + "/remove/" + id;
  return this.del(url);
};


module.exports = AO;
